packages/plugins/response-cache/src/plugin.ts

```diff
@@ -441,42 +443,79 @@ export function useResponseCache<PluginContext extends Record<string, any> = {}>
         );
       }
 
-      return {
-        onExecuteDone({ result, setResult }) {
-          if (isAsyncIterable(result)) {
-            // eslint-disable-next-line no-console
-            console.warn(
-              '[useResponseCache] AsyncIterable returned from execute is currently unsupported.',
-            );
-            return;
-          }
+      async function maybeCacheResult(
+        result: ExecutionResult,
+        setResult: (newResult: ExecutionResult) => void,
+      ) {
+        const processedResult = processResult(result) as ResponseCacheExecutionResult;
 
-          const processedResult = processResult(result) as ResponseCacheExecutionResult;
+        if (skip) {
+          return;
+        }
 
-          if (skip) {
-            return;
-          }
+        if (!shouldCacheResult({ cacheKey, result: processedResult })) {
+          return;
+        }
 
-          if (!shouldCacheResult({ cacheKey, result: processedResult })) {
-            return;
+        // we only use the global ttl if no currentTtl has been determined.
+        const finalTtl = currentTtl ?? globalTtl;
+
+        if (finalTtl === 0) {
+          if (includeExtensionMetadata) {
+            setResult(resultWithMetadata(processedResult, { hit: false, didCache: false }));
           }
+          return;
+        }
 
-          // we only use the global ttl if no currentTtl has been determined.
-          const finalTtl = currentTtl ?? globalTtl;
+        cache.set(cacheKey, processedResult, identifier.values(), finalTtl);
+        if (includeExtensionMetadata) {
+          setResult(
+            resultWithMetadata(processedResult, { hit: false, didCache: true, ttl: finalTtl }),
+          );
+        }
+      }
 
-          if (finalTtl === 0) {
-            if (includeExtensionMetadata) {
-              setResult(resultWithMetadata(processedResult, { hit: false, didCache: false }));
-            }
+      return {
+        onExecuteDone(payload) {
+          if (!isAsyncIterable(payload.result)) {
+            maybeCacheResult(payload.result, payload.setResult);
             return;
           }
 
-          cache.set(cacheKey, processedResult, identifier.values(), finalTtl);
-          if (includeExtensionMetadata) {
-            setResult(
-              resultWithMetadata(processedResult, { hit: false, didCache: true, ttl: finalTtl }),
-            );
-          }
+          // When the result is an AsyncIterable, it means the query is using @defer or @stream.
+          // This means we have to build the final result by merging the incremental results.
+          // The merged result is then used to know if we should cache it and to calculate the ttl.
+          let result: ExecutionResult = {};
+          return {
+            onNext(payload) {
+              const { data, errors, extensions, incremental, hasNext } =
```

Comment:
I'm not happy with this, if someone knows how to make typescript happy by discriminating the union type... it seems `if(payload.result.incremental)` is not working, while I thought it was the way to do it.